azurelinuxagent/ga/update.py

```diff
@@ -986,6 +1009,55 @@ def _ensure_firewall_rules_persisted(dst_ip):
             message=msg,
             log_event=False)
 
+    def __upgrade_agent_if_permitted(self):
+        """
+        Check every 4hrs for a Hotfix Upgrade and 24 hours for a Normal upgrade and upgrade the agent if available.
+        raises: ExitException when a new upgrade is available in the relevant time window, else returns
+        """
+
+        def get_next_process_time(last_val, frequency):
+            return now if last_val is None else last_val + frequency
+
+        now = time.time()
+        next_hotfix_time = get_next_process_time(self._last_hotfix_upgrade_time, conf.get_hotfix_upgrade_frequency())
+        next_normal_time = get_next_process_time(self._last_normal_upgrade_time, conf.get_normal_upgrade_frequency())
+
+        # Not permitted to update yet for any of the AgentUpgradeModes
+        if next_hotfix_time > now and next_normal_time > now:
+            return
+
+        # Update the last upgrade check time even if no new agent is available for upgrade
+        self._last_hotfix_upgrade_time = now if next_hotfix_time <= now else self._last_hotfix_upgrade_time
+        self._last_normal_upgrade_time = now if next_normal_time <= now else self._last_normal_upgrade_time
+
+        available_agent = self.get_latest_agent()
+        if available_agent is None:
+            logger.verbose("No agent upgrade discovered")
```

Comment:
Guess it is beneficial to include current agent version here in the logs

Reply:
We already have that data in the OSInfo log that we emit at the beginning of service, I particularly don't see any advantage of adding it here too. Plus this is a verbose log, it wont be visible in most of the cases to begin with.